### ISSUE TITLE
Is Engine.io support going to be dropped?

### ISSUE THREAD

**Issue by CodingNemesis** [deepstreamIO/deepstream.io#437]
There is a mention of a plan to drop TCP & engine.io support while going through below post. 

https://github.com/deepstreamIO/deepstream.io/issues/68

We are doing an evaluation of technologies to choose our next real time web platform (currently on a commercial system) and dropping of support for engine.io can be a deal breaker for us as we still need websocket fallbacks in browsers. Can somebody from the deepstream core dev group give more clarity on the plan?

Based on our own analytics data, 97-98% of our users do get connected over Web Sockets, but there are still a few users who get connected over HTTP streaming or polling transports.  The reason is not always their browser capability, but often, the kind of networks they are in (where Web Sockets could end up getting blocked). We do not want to loose 2-3% customers by mandating WebSocket support.

**Comment by WolframHempel** [comment]
@CodingNemesis 

Yes, engine.io support will be dropped. However we have been looking very carefully into the issue and have built realtime solutions for large scale, predominantly financial enterprises for a long time.  In the process we've encountered a number of techniques that ensure that websocket upgrades / connection persistence can be ensured even within reverse nats/proxy, firewall and VPN tunnel scenarios.

We are confident that we are able to provide a solution that will work for any endpoint within any corporate infrastructure, provided it has basic websocket support.

If you have a specific scenario in mind that concerns you in particular please let me know and I'll be happy to detail our approach

**Comment by CodingNemesis** [comment]
I am specifically talking about end users out in the internet wilderness and not corporate network scenarios. We have often had cases where something in the user side of the network like a forward proxy, firewall or some other such software that may be running on the same machine or on some other hop in between (and some other unknown reasons as well) that could end up blocking the Websocket connections to our servers. 

As I said before we are presently using a commercial (self-hosted) solution that also claims to use several tricks to get the ideal path (Websocket) working in most of the scenarios, but we still end up seeing a few HTTP fallback situations. 

In the absence of HTTP fallback mechanisms, will deepstream be able to get the Websocket connection working without requiring the user to do anything on their side? Note that in our case, end users are not tech savvy and cannot be expected to get involved in technical troubleshooting or making changes.

**Comment by CodingNemesis** [comment]
Just to add to that, in our compatibility matrix, we also need to support a few older browsers without websocket support even if the market is small.

I do not think we are still at a place where we can comfortably drop support for Websocket polyfills. I would even say that there would be a lot of other people for whom not having browser fallback would be a deal breaker.

Suggest reconsidering the decision. What about integrating SockJS (instead of engine.io) which exposes a pure web socket endpoint in addition to the browser fall backs.

Deepstream looked like a match for our requirements (along with another project Centrifugo) and I even built a basic NATS messaging connector (which I linked in the open issue for NATS conenctor). But lack of websocket fall backs is probably going to be a deal breaker.

**Comment by WolframHempel** [comment]
I perfectly understand your concerns - we've exhaustively researched the reality of current websocket support and potential fallbacks before making this decision.  The main points were:

- Browser Support:  The predominant browser share without websocket support is [IE 9 and below](http://caniuse.com/#feat=websockets). This accounts for a [total of 0.81%](http://caniuse.com/usage-table) of all browsers.

- Scalability Limits: Hybrid implementations, such as engine.io's HTTP longpolling to Websocket upgrade cannot be used in conjunction with distributed loadbalancers due to their common lack of support for sticky sessions between protocols. This means that with engine.io, deepstream would have been unable to run behind e.g. AWS Elastic Load Balancer etc.

- Cross Application Support. Websockets work in almost all systems and programming languages. This removes the limitation to languages that support engine.io as well as the need for the additional TCP connection deepstream currently offers.

- Long-Polling substitutes. We're working with Alex, the guy that build [uWS](https://github.com/uWebSockets/uWebSockets) on the high performance protocol layer behind deepstream which will be a hybrid of WS and DRP, our proprietary protocol. This combination goes beyond the WS specification and has for instance the potential to re-purpose WS' upgrade request to facilitate polling/long-polling scenarios without the need for HTTP. This way, we'd get the best of both worlds.

Bottom line is: We've exhaustively worked with "comet" techniques for the last ten years and always used a combination of Forever Frame / HTML File / Flash Bridge / Long Polling / Chunked encoded pages in iFrames etc... Deepstream is on the best way to become a performance class of its own and we are confident that our protocol choice will allow it to reach the same coverage than any stack of fallbacks can achieve - at orders of magnitude better performance

**Comment by jdmnd** [comment]
Version 2.0.0 has been released which dropped support for engine.io
